--- cacti-0.8.7d/host.php	2009-05-26 21:18:29.000000000 +0200
+++ plugin087d/host.php	2009-06-01 14:50:09.000000000 +0200
@@ -139,6 +139,7 @@
 		/* ==================================================== */
 
 		db_execute("replace into host_graph (host_id,graph_template_id) values (" . $_POST["id"] . "," . $_POST["graph_template_id"] . ")");
+		api_plugin_hook_function('add_graph_template_to_host', array("host_id" => $_POST["id"], "graph_template_id" => $_POST["graph_template_id"]));
 
 		header("Location: host.php?action=edit&id=" . $_POST["id"]);
 		exit;
