--- cacti-0.8.7d/lib/data_query.php	2009-02-12 03:55:50.000000000 +0100
+++ plugin087d/lib/data_query.php	2009-05-16 11:41:34.000000000 +0200
@@ -56,6 +56,8 @@
 	/* update the poller cache */
 	update_poller_cache_from_query($host_id, $snmp_query_id);
 
+	api_plugin_hook_function('run_data_query', array("host_id" => $host_id, "snmp_query_id" => $snmp_query_id));
+
 	return (isset($result) ? $result : true);
 }
 
